### PR TITLE
feat(process): Naive namespaces

### PR DESCRIPTION
**Pull request by sh0rez**
Now that lists are always unwrapped, we assume all other objects use `ObjectMeta`, so they do accept a namespace, regardless of whether they actually use it.

This allows us to simplify our code, by always setting it. For yet unknown edge-cases, we allow using the `tanka.dev/namespaced` annotation to override this behaviour.

Removed the nsPatch injection method, as it is not needed anymore. Show and Apply will behave exactly the same now.

I decided against implementing a default list for cluster-wide resources, because this seems to make things only marginally more correct in theory, but has no effect on the outcome, but comes at the cost of maintaining yet another static list in our code. Quick analysis suggested there is no official type that won't use ObjectMeta, which makes me pretty confident here.

Fixes https://github.com/grafana/tanka/issues/269
Fixes https://github.com/grafana/tanka/issues/308

### CODE REVIEW

**Comment by captncraig** [comment]
Here's my concern about just adding namespaces to everything without exception:

If I wanted to export into a file tree like `{cluster}/{namespace}/{kind}-{name}.yaml`, I would find cluster roles under the namespace of their environment instead of under `default` or maybe `` like I would expect to find them. 

The kubernetes API won't complain, but things will not necessarily go into the directories I'd want them to.

Of course we could also work around this by having a set of overlays we put over the k8s jsonnet lib that add the annotation to those constructors as well.

File: pkg/process/namespace.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const (
+	// AnnotationNamespaced can be set on any resource to override the decision
+	// whether 'metadata.namespace' is set by Tanka
+	AnnotationNamespaced = MetadataPrefix + "/namespaced"
+)
+
+// Namespace injects the default namespace of the environment into each
+// resources, that does not already define one. AnnotationNamespaced can be used
+// to disable this per resource
+func Namespace(list manifest.List, def string) manifest.List {
```
Comment (captncraig) [inline pkg/process/namespace.go:16]: Perhaps this method could be named a little more descriptively. Maybe `AddNamespace` or `EnsureNamespace`.

Comment (sh0rez) [reply pkg/process/namespace.go:16]: I tried to name the methods inside of process by single verbs, as they are a chain of actions:

- Extract
- Unwrap
- Namespace
- Label
- Filter
- Sort

(Hoping "to namespace something" is actually a verb)

**Comment by hamishforbes** [comment]
How would this affect an 'environment' with resources across namespaces?

I would ideally like to have a setup like
```
environments/cluster1/base/
environments/cluster1/myapp-dev1/
environments/cluster1/myapp-dev2/
environments/cluster2/base/
environments/cluster2/myapp-dev3/
lib/myapp/
lib/base/
```

Where the `myapp` lib is a collection of resources for an instance of our application and `base` is our cluster configuration, ingress, logging, metrics etc.

App environments are all under 1 namespace but the cluster baseline is spread across a few.
i don't really want to have to do `cluster/base-logging`, `cluster/base-metrics` etc as separate tanka environments, conceptually they're the same thing and i want to apply them all at once.

**Comment by sh0rez** [comment]
@hamishforbes this should behave exactly as it did before, at least during apply.

Tanka will only ever set the namespace defined in `spec.json`, if your resources that come from Jsonnet don't already have one. If they do, Tanka won't touch that (it never did).

Only two differences will be there:
- Cluster-wide resources will show a namespace in YAML, but this will always be discarded by kubectl afterwards. This makes our code much simpler however
- Namespaces will already be present during `show` or `export`, so they are actually passed to CD tools like Argo or Flux if needed

**Comment by Duologic** [comment]
I'd rather see Tanka or Kubectl return an error (or at least a warning), expecting the developer/jsonnet to set the namespace instead, depending on Tanka to resolve the namespace promotes negligence imo.

File: docs/docs/namespaces.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+---
+name: Namespaces
+route: /namespaces
+---
+
+# Namespaces
+
+When using Tanka, namespaces are handled slightly different compared to
+`kubectl`, because environments offer more granular control than contexts used
+by `kubectl` do.
```
Comment (malcolmholmes) [inline docs/docs/namespaces.md:10]:
```suggestion
by `kubectl`.
```

Comment (malcolmholmes) [reply docs/docs/namespaces.md:10]: If I remember it right, 'prepositions' are something never to end a sentence with.

File: pkg/kubernetes/manifest/errors.go [hunk 3]
```diff
@@ -3,38 +3,45 @@ package manifest
...
-		msg += fmt.Sprintf(":\n\n%s\n\nPlease check above object.", SampleString(s.Manifest.String()).Indent(2))
+		msg += bluef("\nPlease check below object:\n")
+		msg += SampleString(s.Manifest.String()).Indent(2)
 	}
 
 	return msg
 }
 
+var redf = color.New(color.FgRed, color.Bold, color.Underline).Sprintf
+var yellowf = color.New(color.FgYellow).Sprintf
```
Comment (malcolmholmes) [inline pkg/kubernetes/manifest/errors.go:42]: defined in line 41, but used in line 30?!?

**Comment by hamishforbes** [comment]
Would it be an option for this behaviour to be configurable?

I think I would prefer to require namespace to be explicitly set and it would be easier to lint/validate in a CI pipeline if Tanka didn't automatically inject namespaces.

**Comment by sh0rez** [comment]
@hamishforbes There is already discussion going about about this exact thing internally and I am happy to have this debate more broadly. Could you open a separate issue for this?

Generally spoken, in an attempt to have as few knobs as possible, I am not a fan of configurable things. Instead, we should aim for a solution that works better for everybody.

Also you could join our monthly community call (https://docs.google.com/document/d/1mEsc0GxlnwbWAXzbIP7tBb6T5WgAI66_0gIWJzqB93o) this evening and bring this up!

**Pull request merged**